File: orxo/src/main/java/com/stephanhener/android/orxo/OrxoBus.java
package com.stephanhener.android.orxo;


import com.jakewharton.rxrelay.Relay;


import android.support.annotation.IntDef;
import android.util.Log;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.HashMap;
import java.util.Map;

import rx.Scheduler;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;


public abstract class OrxoBus<T> {

    private static final String TAG = "Orxo";
    public static final int EXCEPTION = 0;
    public static final int IGNORE = 1;
    public static final int MULTIPLE = 2;


    @Retention(RetentionPolicy.SOURCE)
    @IntDef({
            EXCEPTION,
            IGNORE,
            MULTIPLE
    })
    public @interface MultiType {
    }

    private Relay<T, T> mRelay;

    private HashMap<Object, HashMap<String, Subscription>> mObjectMap = new HashMap<>();

    public OrxoBus() {
        mRelay = provideRelay();
    }

    protected @MultiType int getMultiType() {
        return IGNORE;
    }


    protected abstract Relay<T, T> provideRelay();


    public void unregister(Object _object) {

        if (_object == null) {
            throw new NullPointerException("Object to unregister can't be null");
        }

        HashMap<String, Subscription> subscriptions = mObjectMap.get(_object);
        if (subscriptions == null) {
            Log.w(TAG, "object has no active subscriptions");
        } else {

            for (Map.Entry<String, Subscription> subscriptionEntry : subscriptions.entrySet()) {
                subscriptionEntry.getValue().unsubscribe();
            }
        }
        mObjectMap.remove(_object);

    }

    public <R> void register(Object _object, Class<R> _eventType, Scheduler _scheduler, Action1<R> _onNext) {

        if (_object == null) {
            throw new NullPointerException("Object to register can't be null");
        }

        HashMap<String, Subscription> subscriptionMap = mObjectMap.get(_object);

        if (subscriptionMap == null) {
            subscriptionMap = new HashMap<>();
            mObjectMap.put(_object, subscriptionMap);
            addSubscription(subscriptionMap, _eventType, _scheduler, _onNext);
            return;
        }

        if (subscriptionMap.containsKey(_eventType.getCanonicalName())) {
            switch (getMultiType()) {

                case EXCEPTION:
                    throw new IllegalStateException("Subscribed twice for the event");
                case IGNORE:
                    //do nothing
                    break;
                case MULTIPLE:
                    addSubscription(subscriptionMap, _eventType, _scheduler, _onNext);
                    break;
            }
        } else {

            addSubscription(subscriptionMap, _eventType, _scheduler, _onNext);
        }
        ;


    }

    private <R> void addSubscription(HashMap<String, Subscription> _subscriptionMap, Class<R> _eventType, Scheduler _scheduler, Action1<R> _onNext) {
        Subscription subscription = mRelay.ofType(_eventType)
                .subscribeOn(Schedulers.io())
                .observeOn(_scheduler)
                .subscribe(_onNext, new Action1<Throwable>() {
                    @Override
                    public void call(Throwable exception) {
                        Log.e("TAG", "error subscribing");
                    }
                });
        _subscriptionMap.put(_eventType.getCanonicalName(), subscription);
    }

    public <R> void register(Object _object, Class<R> _eventType, Action1<R> _onNext) {
        register(_object, _eventType, AndroidSchedulers.mainThread(), _onNext);
    }

    public void send(T _object) {
        mRelay.call(_object);
    }

}

File: orxo/src/main/kotlin/com/tractive/android/orxo/OrxoBus.kt
package com.tractive.android.orxo

import com.jakewharton.rxrelay.Relay
import rx.Scheduler
import rx.Subscription
import rx.functions.Action1
import rx.subscriptions.CompositeSubscription
import java.util.*


abstract class OrxoBus {

    val relay: Relay<Any, Any> by lazy {
        provideRelay()
    }

    private val subscriptions: HashMap<Any, CompositeSubscription> by lazy {
        HashMap<Any, CompositeSubscription>()
    }

    protected abstract fun provideRelay(): Relay<Any, Any>


    fun unregister(_object: Any) {
        subscriptions[_object]?.apply {
            clear()
            subscriptions.remove(_object)
        }
    }

    fun <T> getEvent(_subscriber: Any, _event: Class<T>, _scheduler: Scheduler, _action: Action1<T>) = this@OrxoBus.relay
            .ofType(_event)
            .observeOn(_scheduler)
            .subscribe(_action)
            .register(_subscriber)

    fun send(_subscriber: Any) = relay.call(_subscriber)

    private fun register(_object: Any, subscription: Subscription) {
        subscriptions[_object] = (subscriptions[_object] ?: CompositeSubscription())
                .apply {
                    add(subscription)
                }
    }

    private fun Subscription.register(_subscriber: Any) = register(_subscriber, this)


}

class OrxoHandler(val bus: OrxoBus, private val subscriber: Any, private val scheduler: Scheduler) {

    fun <T> subscribe(_event: Class<T>, _action1: Action1<T>) {
        subscribe(_event, scheduler, _action1)
    }

    fun <T> subscribe(_event: Class<T>, _schedule: Scheduler, _action1: Action1<T>) {
        bus.getEvent(subscriber, _event, _schedule, _action1);
    }
}






